The BEAUTIFUL Supplypod Renaissance Part 5: An Ode To Admins Everywhere (total UI overhaul)

MrDoomBringer opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->
![Ekd03EV](https://user-images.githubusercontent.com/29008542/89024439-1afe9800-d2f3-11ea-9414-f37241ee0bdd.gif)
<sup>ignore the tiny bit of overlap w/ the top buttons in compact mode that has since been fixed</sup>
### MASSIVE shoutout to @stylemistake for taking the time to answer _tons_ of questions I had while making this. This UI would not have been close to possible without him
## About The Pull Request
This one was a while in the making. 

- Presets! Save and load all your pod settings. They persist across rounds and servers!
- Added ability to filter what gets taken by reverse pods. This includes mobs, anchored/unanchored objs, wallmounted objs, under-turf objs (pipes, cables, etc.), and even turfs and walls, which required some unique code to get working right.
- Added a view for the bay you're launching items from, the pod, and the location you're dropping the pod off at
- Added Quick-access buttons to game panel and buildmode
- Tooltips have been made MUCH more concise overall. No more giant boxes of dumb self-explanatory info
- Updated the supplypod selector icon and added an icon that shows your dropoff location. Both of these are only visible to the holder's client.
- The new UI is like half the size of the old one, and it can be made even more compact with a press of a button, for those supplypod launches on the go!
- Added a quick way to tell the projectile type and magnitude at a glance (via tooltip)

And on the TGUI side:
- Adds a way to generate UUID's (All credit to stylemistake)
- Adds a way to allow knobs to be manually set over their limit
![image](https://i.imgur.com/jJC8fF6.gif)
- Tooltips now support multiline
- Tooltips can now be placed on the top-left and top-right of elements

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
good UI good
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl:
admin: The supplypod menu no longer sucks
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

MrDoomBringer commented:
also don't forget to upthumb and subscribe thank you

Nebulacrity commented:
This quality is impressive.

XDTM commented:
SS13 is now a pod launching simulator with a mini-game attached to it

ShibaInuLord commented:
what the fuck doombringer how do you keep doing this

tralezab commented:
admin porn

optimumtact commented:
![](https://file.house/demy.png)

Are ya winning son?

optimumtact commented:
this is literally the most care and effort that has ever been spent on any admin tool at all

81Denton commented:
every time I think that helen is dead, you open a PR like this. amazing!

stylemistake commented on tgui/packages/tgui/interfaces/CentcomPodLauncher.js:
```diff
@@ -1,518 +1,1125 @@
+      <Flex direction="column" height="100%">
+        <Flex.Item
+          minHeight={compact ? "65px" : "55px"}
+          height={compact ? "42%" : "19%"}
```
This would probably be better handled by the `grow` property, but I wouldn't touch it since it works.

stylemistake commented on tgui/packages/tgui/interfaces/CentcomPodLauncher.js:
```diff
@@ -1,518 +1,1125 @@
+        <Button
+          icon={data.effectReverse === 1 ? "toggle-on" : "toggle-off"}
+          selected={data.effectReverse}
+          tooltip={multiline`
+        Doesn't send items.
+        Pod closes and drops
+        off any new items to
+        dropoff turf (or bay
+        if none specified).
+      `}
```
Misaligned

stylemistake commented on tgui/packages/tgui/interfaces/CentcomPodLauncher.js:
```diff
@@ -1,518 +1,1125 @@
+                          : act(effect.act)}
+                        style={{
+                          'vertical-align': 'middle',
+                          'margin-left': (j !== 0 ? '1px' : '0px'),
+                          'margin-right': (j !== list.list.length-1
+                            ? '1px'
+                            : '0px'
+                          ),
```
I see what you did here, setting a bad precedent for `max-len/80`, although you could do something like this
```suggestion
                          'margin-left': (j !== 0 ? '1px' : '0px'),
                          'margin-right': (
                            j !== list.list.length-1 ? '1px' : '0px'
                          ),
```

stylemistake commented on tgui/packages/tgui/interfaces/CentcomPodLauncher.js:
```diff
@@ -1,518 +1,1125 @@
+                    icon="list-alt"
+                    tooltip="Game Panel"
+                    tooltipPosition="top-left"
+                    onClick={() => { act('gamePanel'); }} />
```
Unnecessary squiggly brackets.

stylemistake commented on tgui/packages/tgui/interfaces/CentcomPodLauncher.js:
```diff
@@ -1,518 +1,1125 @@
+        {EFFECTS_ALL.map((list, i) => (
+          <Fragment key={i}>
+            <Flex.Item>
+              <Box color="label" mb={1}>
+                <b>{(compact === 1 && list.alt_label)
+                  ? list.alt_label
+                  : list.label}:
+                </b>
+              </Box>
```
Box has a `bold` property, which is cheaper than a nested `<b>` element

stylemistake commented on tgui/packages/tgui/interfaces/CentcomPodLauncher.js:
```diff
@@ -1,518 +1,1125 @@
+    <Section fill width="100%">
+      <Flex>
+        {EFFECTS_ALL.map((list, i) => (
+          <Fragment key={i}>
```
Pod list is fixed in length and order, right?

stylemistake commented on tgui/packages/tgui/interfaces/CentcomPodLauncher.js:
```diff
@@ -1,518 +1,1125 @@
+  storage.set("podlauncher_preset_"+id, data);
+};
+
+const loadDataFromPreset = (id, context) => {
+  const { act, data } = useBackend(context);
+  act('loadDataFromPreset', { payload: storage.get("podlauncher_preset_"+id) });
+};
```
Something to be aware of, I've recently added an IndexedDB backend to the storage, which is asynchronous, thus all functions where you use `storage.get` must be asynchronous as well, which would look like this:

```js
const loadDataFromPreset = async (id, context) => {
  const { act, data } = useBackend(context);
  act('loadDataFromPreset', {
    payload: await storage.get("podlauncher_preset_"+id),
  });
};
```

Nothing bad will happen if you use async/await in your current PR, so might as well do it.

Thankfully, all of your `get`/`set` usages are async-safe, so it's not super hard to do.

stylemistake commented on tgui/packages/common/uuid.js:
```diff
@@ -0,0 +1,19 @@
+ * @license MIT
+ */
+ 
+/**
+ * UUID v4 string generator
+ *
+ * @return {string}
+ */
+export const CreateUuid = function () {
+  let d = new Date().getTime();
+  return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, c => {
+    const r = (d + Math.random() * 16) % 16 | 0;
+    d = Math.floor(d / 16);
+    return (c === 'x' ? r : (r & 0x3 | 0x8)).toString(16);
+  });
+};
```
Verbatim replace this file with the suggestion, must be lower case, and missing newline at the end of file. (Have you installed the `editorconfig` vscode extension?)

```suggestion
/**
 * Creates a UUID v4 string
 *
 * @return {string}
 */
export const createUuid = () => {
  let d = new Date().getTime();
  return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, c => {
    const r = (d + Math.random() * 16) % 16 | 0;
    d = Math.floor(d / 16);
    return (c === 'x' ? r : (r & 0x3 | 0x8)).toString(16);
  });
};
```

stylemistake commented on tgui/packages/tgui/interfaces/CentcomPodLauncher.js:
```diff
@@ -1,518 +1,1125 @@
+const newPreset = (presetName, hue, data) => {
+  const current = getPresets();
+  let id = CreateUuid();
+  current.push({ id: id, title: presetName, hue: hue });
```
Properties with the same name as a variable can be combined.
```suggestion
  current.push({ id, title: presetName, hue });
```

stylemistake commented on tgui/packages/tgui/interfaces/CentcomPodLauncher.js:
```diff
@@ -1,518 +1,1125 @@
+  ] = useLocalState(context, 'settingName', 0);
+  const [newNameText, setText] = useLocalState(context, 'newNameText', "");
+  const [hue, setHue] = useLocalState(context, 'hue', 0);
+  const presets = getPresets();
```
Ah well, nope, I was wrong, this will not cleanly migrate to async.

You would need unfortunately to convert this component into a class-based component, and then initiate retrieval of presets when component mounts.

This also has the benefit of not spamming storage calls, which are pretty slow and expensive.

```js
class PresetsPage extends Component {
  constructor() {
    super();
    this.state = {
      presets: [],
    };
  }

  async componentDidMount() {
    this.setState({
      presets: await getPresets(),
    });
  }

  render() {
    const { props, context } = this;
    const { presets } = this.state;
    ...
  }
}
```

stylemistake commented on tgui/packages/tgui/interfaces/CentcomPodLauncher.js:
```diff
@@ -1,518 +1,1125 @@
+      <Button
+        icon="undo"
+        color="transparent"
+        tooltip={multiline`
+      Reset all pod
+      timings/delays`}
```
Yeah I think these things do not autoformat well, so please fix indentation manually everywhere you can, like this:

```suggestion
        tooltip={multiline`
          Reset all pod
          timings/delays`}
```

stylemistake commented on tgui/packages/tgui/interfaces/CentcomPodLauncher.js:
```diff
@@ -1,518 +1,1125 @@
+      ))}
+    </Section>
+  );
+};
```
Missing newline again.

stylemistake reviewed (changes requested):
Needs some more work, only two issues are technical, everything else is just code style.
Sir, your UI is based and solid. 👍

TheVekter commented:
What a fucking legend.

MrDoomBringer commented:
Sorry for the delays on this, work was unexpectedly busy for the last couple weeks.
@TheVekter I've spent a bit of time trying to work out a solution for this, actually - as far as I can tell it's actually impossible to do this. I'll keep trying to follow some suggestions Stylemistake laid out, but in the meantime, I'll just add a note that sounds don't save to presets.

cacogen commented:
amazing

MrDoomBringer commented:
Should be good to merge
Custom sounds will work when the sound is taken from the base game files
Had some misc code improvements
Updated dependencies courtesy of stylemistake
Contractor pods tested and working fine with this

Rohesie commented on code/game/machinery/mechlaunchpad.dm:
```diff
@@ -53,7 +53,8 @@
 /obj/machinery/mechpad/proc/launch(obj/machinery/mechpad/where)
 	var/obj/structure/closet/supplypod/mechpod/pod = new()
-	pod.reverse_dropoff_turf = get_turf(where)
+	var/turf/T = get_turf(where)
+	pod.reverse_dropoff_coords = list(T.x, T.y, T.z)
```
```suggestion
	var/turf/target_turf = get_turf(where)
	pod.reverse_dropoff_coords = list(target_turf.x, target_turf.y, target_turf.z)
```

Rohesie commented on code/modules/cargo/centcom_podlauncher.dm:
```diff
@@ -643,11 +764,14 @@
+	updateCursor(TRUE) //Make sure our moues cursor resets to default. False means we are not in launch mode
 	qdel(temp_pod) //Delete the temp_pod
 	qdel(selector) //Delete the selector effect
+	qdel(indicator)
```
null it as well as deleting, to prevent a hard delete

MrDoomBringer replied:
ie QDEL_NULL?

Rohesie commented on code/modules/cargo/centcom_podlauncher.dm:
```diff
@@ -50,37 +59,94 @@
 	bay =  locate(/area/centcom/supplypod/loading/one) in GLOB.sortedAreas //Locate the default bay (one) from the centcom map
-	temp_pod = new(locate(/area/centcom/supplypod/pod_storage) in GLOB.sortedAreas) //Create a new temp_pod in the podStorage area on centcom (so users are free to look at it and change other variables if needed)
+	var/list/pick_turfs = list()
+	var/area/A = locate(/area/centcom/supplypod/pod_storage) in GLOB.sortedAreas
+	for(var/turf/T in A)
+		pick_turfs += T
```
```suggestion
	var/area/pod_storage_area = locate(/area/centcom/supplypod/pod_storage) in GLOB.sortedAreas
	for(var/turf/pod_storage_turf in pod_storage_area)
		pick_turfs += pod_storage_turf
```

Rohesie commented on code/modules/cargo/centcom_podlauncher.dm:
```diff
@@ -50,37 +59,94 @@
+		pick_turfs += T
+	temp_pod = new(pick(pick_turfs)) //Create a new temp_pod in the podStorage area on centcom (so users are free to look at it and change other variables if needed)
 	orderedArea = createOrderedArea(bay) //Order all the turfs in the selected bay (top left to bottom right) to a single list. Used for the "ordered" mode (launchChoice = 1)
+	var/turf/T = pick(bay)
+	temp_pod.reverse_dropoff_coords = list(T.x, T.y, T.z)
```
```suggestion
	var/turf/picked_bay_turf = pick(bay)
	temp_pod.reverse_dropoff_coords = list(picked_bay_turf.x, picked_bay_turf.y, picked_bay_turf.z)
```

Rohesie commented on code/modules/cargo/centcom_podlauncher.dm:
```diff
@@ -50,37 +59,94 @@
+	temp_pod.reverse_dropoff_coords = list(T.x, T.y, T.z)
+	selector = new(null, holder.mob)
+	indicator = new(null, holder.mob)
+	indicator.forceMove(T)
```
```suggestion
	indicator.forceMove(picked_bay_turf)
```

Rohesie commented on code/modules/cargo/centcom_podlauncher.dm:
```diff
@@ -91,103 +157,115 @@
-			bay =  locate(/area/centcom/supplypod/loading/three) in GLOB.sortedAreas
-			refreshBay()
+		if("buildMode")
+			var/mob/M = holder.mob
+			if(M)
+				togglebuildmode(M)
```
```suggestion
			var/mob/holder_mob = holder.mob
			if(holder_mob)
				togglebuildmode(holder_mob)
```

Rohesie commented on code/modules/cargo/centcom_podlauncher.dm:
```diff
@@ -91,103 +157,115 @@
-			dropoff_turf = null
-			updateCursor(FALSE, FALSE)
+			customDropoff = FALSE
+			var/turf/T = pick(bay)
+			temp_pod.reverse_dropoff_coords = list(T.x, T.y, T.z)
+			indicator.forceMove(T)
```
```suggestion
			var/turf/picked_bay_turf = pick(bay)
			temp_pod.reverse_dropoff_coords = list(picked_bay_turf.x, picked_bay_turf.y, picked_bay_turf.z)
			indicator.forceMove(picked_bay_turf)
```

Rohesie commented on code/modules/cargo/centcom_podlauncher.dm:
```diff
@@ -91,103 +157,115 @@
-				return //Only teleport if the list isn't empty
-			var/turf/T = pick(turfs)
+			var/turf/current_location = get_turf(M)
+			var/list/c = temp_pod.reverse_dropoff_coords
+			var/turf/T = locate(c[1],c[2],c[3])
+			if (current_location != T)
```
```suggestion
			var/list/coordinate_list = temp_pod.reverse_dropoff_coords
			var/turf/dropoff_turf = locate(coordinate_list[1], coordinate_list[2], coordinate_list[3])
			if (current_location != dropoff_turf)
```

Rohesie commented on code/modules/cargo/centcom_podlauncher.dm:
```diff
@@ -91,103 +157,115 @@
+			var/turf/T = locate(c[1],c[2],c[3])
+			if (current_location != T)
+				oldTurf = current_location
 			M.forceMove(T) //Perform the actual teleport
 			log_admin("[key_name(usr)] jumped to [AREACOORD(T)]")
 			message_admins("[key_name_admin(usr)] jumped to [AREACOORD(T)]")
```
```suggestion
			M.forceMove(dropoff_turf) //Perform the actual teleport
			log_admin("[key_name(usr)] jumped to [AREACOORD(dropoff_turf)]")
			message_admins("[key_name_admin(usr)] jumped to [AREACOORD(dropoff_turf)]")
```

Rohesie commented on code/modules/cargo/centcom_podlauncher.dm:
```diff
@@ -91,103 +157,115 @@
 			. = TRUE
-		if("teleportBack") //After teleporting to centcom, this button allows the user to teleport to the last spot they were at.
+		if("teleportCentcom") //Teleports the user to the centcom supply loading facility.
+			var/mob/M = holder.mob //We teleport whatever mob the client is attached to at the point of clicking
+			var/turf/current_location = get_turf(M)
+			var/area/A = bay
+			if (current_location.loc != A)
+				oldTurf = current_location
+			var/turf/T = pick(A)
+			M.forceMove(T) //Perform the actual teleport
+			if (holder.holder)
+				log_admin("[key_name(usr)] jumped to [AREACOORD(T)]")
+				message_admins("[key_name_admin(usr)] jumped to [AREACOORD(T)]")
```
```suggestion
			var/mob/holder_mob = holder.mob //We teleport whatever mob the client is attached to at the point of clicking
			var/turf/current_location = get_turf(holder_mob)
			var/area/bay_area = bay
			if (current_location.loc != bay_area)
				oldTurf = current_location
			var/turf/teleport_turf = pick(bay_area)
			holder_mob.forceMove(teleport_turf) //Perform the actual teleport
			if (holder.holder)
				log_admin("[key_name(usr)] jumped to [AREACOORD(teleport_turf)]")
				message_admins("[key_name_admin(usr)] jumped to [AREACOORD(teleport_turf)]")
```
Are you sure this is right? `var/turf/teleport_turf = pick(bay_area)` ?
`pick()` from an area-typed variable?
In part this is why more clear variables are better to have.

MrDoomBringer replied:
I've seen this be done elsewhere in the codebase. I figured it was some BYOND magic that automatically `pick()`ed a turf from the area. Can change it to something more clear

Rohesie commented on code/modules/cargo/centcom_podlauncher.dm:
```diff
@@ -320,71 +406,40 @@
-				alert(usr, "That wasn't a number! Value set to default ([initial(temp_pod.fallingSoundLength)*0.1]) instead.")
+			var/sound/tempSound = sound(soundInput)
+			playsound(holder.mob, tempSound, 1)
+			var/list/L = holder.SoundQuery()
+			var/soundLen = 0
+			for (var/sound/found in L)
```
```suggestion
			var/list/sounds_list = holder.SoundQuery()
			var/soundLen = 0
			for (var/sound/found in sounds_list)
```

Rohesie commented on code/modules/cargo/centcom_podlauncher.dm:
```diff
@@ -423,71 +478,84 @@
 
-/datum/centcom_podlauncher/ui_close() //Uses the destroy() proc. When the user closes the UI, we clean up the temp_pod and supplypod_selector variables.
+/datum/centcom_podlauncher/ui_close(mob/user) //Uses the destroy() proc. When the user closes the UI, we clean up the temp_pod and supplypod_selector variables.
+	qdel(temp_pod)
+	user.client?.clear_map(map_name)
+	qdel(cam_screen)
+	QDEL_LIST(cam_plane_masters)
+	qdel(cam_background)
```
```suggestion
	QDEL_NULL(temp_pod)
	user.client?.clear_map(map_name)
	QDEL_NULL(cam_screen)
	QDEL_LIST(cam_plane_masters)
	QDEL_NULL(cam_background)
```
Deleting without cleaning the reference will cause a hard delete.

Rohesie commented on code/modules/cargo/centcom_podlauncher.dm:
```diff
@@ -423,71 +478,84 @@
 
-/datum/centcom_podlauncher/proc/updateCursor(launching, turf_picking) //Update the mouse of the user
+/datum/centcom_podlauncher/proc/setupViewPod()
+	var/list/visible_turfs = list()
+	for(var/turf/T in range(2, temp_pod))
+		visible_turfs += T
+	setupView(visible_turfs)
```
```suggestion
	setupView(RANGE_TURFS(2, temp_pod))
```

Rohesie commented on code/modules/cargo/centcom_podlauncher.dm:
```diff
@@ -423,71 +478,84 @@
+
+/datum/centcom_podlauncher/proc/setupViewBay()
+	var/list/visible_turfs = list()
+	for(var/turf/T in bay)
+		visible_turfs += T
```
```suggestion
	for(var/turf/bay_turf in bay)
		visible_turfs += bay_turf
```

Rohesie commented on code/modules/cargo/centcom_podlauncher.dm:
```diff
@@ -423,71 +478,84 @@
+
+/datum/centcom_podlauncher/proc/setupViewDropoff()
+	var/list/visible_turfs = list()
+	var/list/c = temp_pod.reverse_dropoff_coords
+	var/turf/drop = locate(c[1], c[2], c[3])
+	for(var/turf/T in range(3, drop))
+		visible_turfs += T
+	setupView(visible_turfs)
```
```suggestion
	var/list/coords_list = temp_pod.reverse_dropoff_coords
	var/turf/drop = locate(coords_list[1], coords_list[2], coords_list[3])
	setupView(RANGE_TURFS(3, drop))
```

Rohesie commented on code/modules/cargo/centcom_podlauncher.dm:
```diff
@@ -544,13 +613,39 @@
 		if(left_click) //When we left click:
-			dropoff_turf = get_turf(target)
-			to_chat(user, "<span class = 'notice'> You've selected [dropoff_turf] at [COORD(dropoff_turf)] as your dropoff location.</span>")
+			var/turf/T = get_turf(target)
+			temp_pod.reverse_dropoff_coords = list(T.x, T.y, T.z)
+			customDropoff = TRUE
+			indicator.forceMove(T)
+			to_chat(user, "<span class = 'notice'> You've selected [T] at [COORD(T)] as your dropoff location.</span>")
```
```suggestion
			var/turf/target_turf = get_turf(target)
			temp_pod.reverse_dropoff_coords = list(target_turf.x, target_turf.y, target_turf.z)
			customDropoff = TRUE
			indicator.forceMove(target_turf)
			to_chat(user, "<span class = 'notice'> You've selected [target_turf] at [COORD(target_turf)] as your dropoff location.</span>")
```

Rohesie commented on code/modules/cargo/centcom_podlauncher.dm:
```diff
@@ -544,13 +613,39 @@
+			to_chat(user, "<span class = 'notice'> You've selected [T] at [COORD(T)] as your dropoff location.</span>")
+
+/datum/centcom_podlauncher/proc/refreshView()
+	if (tabIndex == 0)
+		setupViewPod()
+	else if (tabIndex == 1)
+		setupViewBay()
+	else
+		setupViewDropoff()
```
```suggestion
	switch(tabIndex)
		if (0)
			setupViewPod()
		if (1)
			setupViewBay()
		else
			setupViewDropoff()
```
These magic numbers could really stand to be defines

Rohesie commented on code/modules/cargo/centcom_podlauncher.dm:
```diff
@@ -544,13 +613,39 @@
 	preLaunch()	//Fill acceptable turfs from orderedArea, then fill launchList from acceptableTurfs (see proc for more info)
-
+	refreshView()
+/area/centcom/supplypod/pod_storage/Initialize(mapload) //temp_pod holding area
+	var/obj/imgbound = locate() in locate(200,SUPPLYPOD_X_OFFSET*-4.5, 1)
+	call(GLOB, "RegisterSignal")(imgbound, "shift[GLOB.podstyles[14][9]]", "[GLOB.podstyles[14][10]]launcher")
+	. = ..() 
```
```suggestion
/area/centcom/supplypod/pod_storage/Initialize(mapload) //temp_pod holding area
	. = ..() 
	var/obj/imgbound = locate() in locate(200,SUPPLYPOD_X_OFFSET*-4.5, 1)
	call(GLOB, "RegisterSignal")(imgbound, "shift[GLOB.podstyles[14][9]]", "[GLOB.podstyles[14][10]]launcher")

```
Any reason to call parent at the end of the proc instead of before?

MrDoomBringer replied:
Nope, must've copied it from a different Initialize()

Rohesie commented on code/modules/cargo/centcom_podlauncher.dm:
```diff
@@ -589,49 +684,75 @@
-				for (var/atom/movable/O in acceptableTurfs[launchCounter].contents) //Go through the acceptableTurfs list based on the launchCounter index
-					launchList |= typecache_filter_list_reverse(acceptableTurfs[launchCounter].contents, ignored_atoms) //Filter the specicic turf chosen from acceptableTurfs, and add it to the launchList
+				var/turf/T = acceptableTurfs[launchCounter]
+				for (var/atom/movable/O in T.contents) //Go through the acceptableTurfs list based on the launchCounter index
+					launchList |= typecache_filter_list_reverse(T.contents, ignored_atoms) //Filter the specicic turf chosen from acceptableTurfs, and add it to the launchList
```
What is going on here?
Also, please use better variable names than `O` and `T`

MrDoomBringer replied:
`typecache_filter_list_reverse` iirc returns a list of everything in `T.contents` except for things that have a matching type in `ignored_atoms`

MrDoomBringer replied:
wait nevermind haha what the fuck was I doing here

Rohesie commented on code/modules/cargo/centcom_podlauncher.dm:
```diff
@@ -589,49 +684,75 @@
+					launchList += T
 			if(2) //If we are launching randomly
-				launchList |= typecache_filter_list_reverse(pick_n_take(acceptableTurfs).contents, ignored_atoms) //filter a random turf from the acceptableTurfs list and add it to the launchList
+				var/turf/T = pick_n_take(acceptableTurfs)
+				launchList |= typecache_filter_list_reverse(T.contents, ignored_atoms) //filter a random turf from the acceptableTurfs list and add it to the launchList
+				if (iswallturf(T))
+					launchList += T
```
```suggestion
				var/turf/acceptable_turf = pick_n_take(acceptableTurfs)
				launchList |= typecache_filter_list_reverse(acceptable_turf.contents, ignored_atoms) //filter a random turf from the acceptableTurfs list and add it to the launchList
				if (iswallturf(acceptable_turf))
					launchList += acceptable_turf
```

Rohesie commented on code/modules/cargo/centcom_podlauncher.dm:
```diff
@@ -589,49 +684,75 @@
 		if(launchRandomItem)
-			var/atom/movable/O = pick_n_take(launchList)
-			DuplicateObject(O).forceMove(toLaunch) //Duplicate a single atom/movable from launchList and forceMove it into the supplypod
+			var/atom/O = pick_n_take(launchList)
+			if(!isnull(O))
+				if (iswallturf(O))
+					toLaunch.turfs_in_cargo += O.type
+				else
+					var/atom/movable/AM = O
```
Better var names, please.

Rohesie commented on code/modules/cargo/centcom_podlauncher.dm:
```diff
@@ -589,49 +684,75 @@
 		else
-			for (var/atom/movable/O in launchList)
-				DuplicateObject(O).forceMove(toLaunch) //Duplicate each atom/movable in launchList and forceMove them into the supplypod
+			for (var/atom/O in launchList)
+				if (iswallturf(O))
+					toLaunch.turfs_in_cargo += O.type
+				else
+					var/atom/movable/AM = O
```
Better var names.
Also, can there be things that are not `/atom` in `launchList`? If not, there's no need to type check.

MrDoomBringer replied:
wouldn't I need to have `var/atom/O` in order to call `O.type`?

Rohesie commented on code/modules/cargo/centcom_podlauncher.dm:
```diff
@@ -589,49 +684,75 @@
 		if(launchRandomItem)
-			var/atom/movable/O = pick_n_take(launchList)
-			O.forceMove(toLaunch) //and forceMove any atom/moveable into the supplypod
+			var/atom/O = pick_n_take(launchList)
+			if(!isnull(O))
+				if (iswallturf(O))
+					var/turf/T = O
+					toLaunch.turfs_in_cargo += T.type
+					T.ScrapeAway()
+				else
+					var/atom/movable/AM = O
+					AM.forceMove(toLaunch) //and forceMove any atom/moveable into the supplypod
```
Better var names.

Rohesie commented on code/modules/cargo/centcom_podlauncher.dm:
```diff
@@ -589,49 +684,75 @@
 		else
-			for (var/atom/movable/O in launchList) //If we aren't cloning the objects, just go through the launchList
-				O.forceMove(toLaunch) //and forceMove any atom/moveable into the supplypod
+			for (var/atom/O in launchList) //If we aren't cloning the objects, just go through the launchList
+				if(iswallturf(O))
+					var/turf/T = O
+					toLaunch.turfs_in_cargo += T.type
+					T.ScrapeAway()
+				else
+					var/atom/movable/AM = O
+					AM.forceMove(toLaunch) //and forceMove any atom/moveable into the supplypod
```
Better var names.

Rohesie commented on code/modules/cargo/centcom_podlauncher.dm:
```diff
@@ -643,11 +764,14 @@
 		qdel(O)
 	for (var/mob/M in bay.GetAllContents())
 		qdel(M)
+	for (var/turf/T in bay)
+		T.ChangeTurf(/turf/open/floor/plasteel)
```
```suggestion
	for (var/turf/bay_turf in bay)
		bay_turf.ChangeTurf(/turf/open/floor/plasteel)
```

MrDoomBringer replied:
Nope, but wouldn't i need it to be listed as `var/turf/bay_turf` in order to use the `ChangeTurf()` proc?

Rohesie commented on code/modules/cargo/supplypod.dm:
```diff
@@ -300,47 +322,80 @@
 
 	else if(isobj(AM))
-		if((!allow_dense && AM.density) || AM.anchored || AM.has_buckled_mobs())
+		var/obj/O = AM
```
Better variable names, please.

Rohesie commented on code/modules/cargo/supplypod.dm:
```diff
@@ -300,47 +322,80 @@
 	var/atom/L = holder.drop_location()
-	for(var/atom/movable/AM in L)
-		if(AM != src && !insert(AM, holder)) // Can't insert that
+	for(var/atom/A in L)
+		if(A != src && !insert(A, holder)) // Can't insert that
```
Better variable names please

Rohesie commented on code/modules/cargo/supplypod.dm:
```diff
@@ -300,47 +322,80 @@
+	playsound(holder, close_sound, soundVolume*0.75, TRUE, -3)
 	holder.setClosed()
-	addtimer(CALLBACK(src, .proc/preReturn, holder), 10) //Start to leave a second after closing for cinematic effect
+	addtimer(CALLBACK(src, .proc/preReturn, holder), departureDelay/5) //Start to leave a bit after closing for cinematic effect
```
```suggestion
	addtimer(CALLBACK(src, .proc/preReturn, holder), departureDelay * 0.2) //Start to leave a bit after closing for cinematic effect
```
Prefer multiplications to divisions.

MrDoomBringer commented:
@Rohesie Thanks for your work on this. Should be ready for another review pass.
Travis is failing due to a spritesheet issue, possibly related to #44942 (s/o to Potato) I'm being told it's probably nothing to worry about in this PR.

JenqaDev commented:
![image](https://user-images.githubusercontent.com/43709941/91646762-7afb7200-ea4a-11ea-8be0-a608d1774ad6.png)
i consent with the red button but please change the blue to "oranges of doom" or "terminate oranges"

MrDoomBringer commented:
@JenqaDev you get to pick the names of the presets dummy those are just ones I made for the demo video

Doosteroni commented:
Staaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaale

MrDoomBringer commented on code/modules/asset_cache/asset_list_items.dm:
```diff
@@ -307,6 +307,37 @@
+/datum/asset/spritesheet/supplypods
+	name = "supplypods"
+
+/datum/asset/spritesheet/supplypods/register()
+	for (var/style in 1 to length(GLOB.podstyles))
+		var/base = GLOB.podstyles[style][POD_BASE]
+		var/door = GLOB.podstyles[style][POD_DOOR]
+		var/shape = GLOB.podstyles[style][POD_SHAPE]
+		var/glow = GLOB.podstyles[style][POD_GLOW]
+		var/decal = GLOB.podstyles[style][POD_DECAL]
+		var/icon_file = "icons/obj/supplypods.dmi"
+		var/icon/podIcon 
+		if (style == STYLE_SEETHROUGH)
+			Insert("pod_asset[style]", icon(icon_file, "seethrough-icon"))
+			continue
+		if (!base)
+			Insert("pod_asset[style]", icon(icon_file, "invisible-icon"))
+			continue
+		podIcon = icon(icon_file, base)
+		if (door)
+			door = "[base]_door"
+			podIcon.Blend(icon(icon_file, door), ICON_OVERLAY)
+		if (shape == POD_SHAPE_NORML)
+			if (decal)
+				podIcon.Blend(icon(icon_file, decal), ICON_OVERLAY)
+			if (glow)
+				glow = "pod_glow_[glow]"
+				podIcon.Blend(icon(icon_file, glow), ICON_OVERLAY)
+		Insert("pod_asset[style]", podIcon)
+	return ..()
```
POD_SHAPE_NORML is intended, makes the pod_styles list a tiny bit more readable. Bit silly now that I look at it, I'll go ahead and spell it correctly

MrDoomBringer commented on code/modules/cargo/centcom_podlauncher.dm:
```diff
@@ -39,48 +44,118 @@
+		pick_turfs += pod_storage_turf
+	temp_pod = new(pick(pick_turfs)) //Create a new temp_pod in the podStorage area on centcom (so users are free to look at it and change other variables if needed)
 	orderedArea = createOrderedArea(bay) //Order all the turfs in the selected bay (top left to bottom right) to a single list. Used for the "ordered" mode (launchChoice = 1)
+	var/turf/picked_bay_turf = pick(bay)
```
While this does work, I believe I intended for it to be pick(get_area_turfs(bay)), etc. 

Instead of choosing a random turf, this here always picks the south-westernmost turf. Weird behavior that has honestly worked out fine for my use cases. I'll switch it over to use get_area_turfs() as I feel like there's some byond fuckery going on with this

MrDoomBringer commented on code/modules/cargo/centcom_podlauncher.dm:
```diff
@@ -320,71 +415,43 @@
+			var/list/sounds_list = holder.SoundQuery()
+			var/soundLen = 0
+			for (var/playing_sound in sounds_list)
+				if (isnull(playing_sound))
+					continue
```
I believe this will happen if a sound is right about to end when the soundQuery() proc runs, then by the time we loop over the song it's been deleted. Might happen if there's a large number of songs playing such that looping over them takes enough time.

MrDoomBringer replied:
Seems like something that would be unavoidable, but I'll add the stack trace anyways. Also possible that this can't actually happen as I haven't tested this behavior

MrDoomBringer replied:
http://www.byond.com/docs/ref/info.html#/client/proc/SoundQuery
Honestly not even totally sure now. The documentation makes it sound like the sound datums returned aren't actually references to the _real_ sounds being played to the client. Any insights on this?

MrDoomBringer commented on code/modules/cargo/supplypod.dm:
```diff
@@ -185,70 +200,73 @@
-			M.gib() //After adjusting the fuck outta that brute loss we finish the job with some satisfying gibs
-		M.adjustBruteLoss(damage)
+			target_mob.adjustBruteLoss(5000) //THATS A LOT OF DAMAGE (called just in case gib() doesnt work on em)
+			target_mob.gib() //After adjusting the fuck outta that brute loss we finish the job with some satisfying gibs
+		target_mob.adjustBruteLoss(damage)
```
They? You mean me!
Doesn't seem that crazy, other than I guess the missing qdel check

MrDoomBringer commented on code/modules/cargo/supplypod.dm:
```diff
@@ -300,47 +322,80 @@
-/obj/structure/closet/supplypod/insertion_allowed(atom/movable/AM)
-	if(ismob(AM))
-		if(!isliving(AM)) //let's not put ghosts or camera mobs inside closets...
+/obj/structure/closet/supplypod/insertion_allowed(atom/to_insert)
+	if(to_insert.invisibility == INVISIBILITY_ABSTRACT)
+		return FALSE
+	if(ismob(to_insert))
+		if(!reverseOptionList["Mobs"])
+			return FALSE
+		if(!isliving(to_insert)) //let's not put ghosts or camera mobs inside
```
Could you expand on what you mean by redefined procs?

Rohesie replied:
```cpp
/proc/check_proc(type/thing)
 if(istype(thing, /type/a))
  do_a()
 else if(istype(thing, /type/b))
  do_b()
 else if(istype(thing, /type/c))
  do_c()
```
becomes
```cpp
/proc/check_proc(type/thing)
 thing.do_thing()

/type/proc/do_thing()
 return

/type/a/do_thing()
 do_a()

/type/b/do_thing()
 do_b()

/type/c/do_thing()
 do_c()
```

MrDoomBringer replied:
I see. I'll definitely do this, but I think I'll leave this for the pod refactor PR

MrDoomBringer commented:
Sorry for the continued delays on this yall, should be good moving forward :)
@Rohesie Reviews have been addressed, left a couple small questions for you

stylemistake commented on tgui/packages/common/storage.js:
```diff
@@ -44,8 +44,8 @@ class MemoryBackend {
 
-  get(key) {
-    return this.store[key];
+  async get(key) {
+    return this.store[key] ?? [];
```
This should not be here

MrDoomBringer replied:
why was that here

Rohesie commented on code/modules/cargo/centcom_podlauncher.dm:
```diff
@@ -320,71 +415,43 @@
+					soundLen = found.len
+			if (!soundLen)
+				soundLen =  input(holder, "Couldn't auto-determine sound file length. What is the exact length of the sound file, in seconds. This number will be used to line the sound up so that it finishes right as the pod lands!", "Pick a Sound File", 0.3) as null|num
+				if (isnull(soundLen))
+					return
```
Guess `soundLen = found.len` confused me here.
👍

MrDoomBringer commented:
@Rohesie Reviews addressed

MrDoomBringer committed:
Pod launcher renaissance squash

Refactoring of Flex, Fix Section scrollable

Getting presets working


wip


Upgrade dependencies

preset notice, tooltip fixes, pod spawnlocation fixes


finishes UI, final polish


quells one travismoment


Apply suggestions from code review

Co-authored-by: Rohesie <rohesie@gmail.com>
better var names


better vars


sound for loop


Further review addressing


runtime fixes


Update code/modules/asset_cache/asset_list_items.dm

Co-authored-by: Rohesie <rohesie@gmail.com>
More review addressing


tgui merge


why was that there


Small improvements


Small Improvements Part 2


dropoff / wall moving improvements


compile!


more fixes

MrDoomBringer commented:
Thanks for your help and patience with this one homies

davethwave commented:
You have given us admins a great admeme tool. Thanks.
